### ISSUE TITLE
Bad error in processing the message java.lang.NullPointerException - route2Subscribers

### ISSUE THREAD

**Issue by kfowlks** [moquette-io/moquette#160]
```
ERROR NettyMQTTHandler  - Bad error in processing the message
java.lang.NullPointerException
    at io.moquette.spi.impl.ProtocolProcessor.route2Subscribers(ProtocolProcessor.java:460)
    at io.moquette.spi.impl.ProtocolProcessor.processPublish(ProtocolProcessor.java:310)
    at io.moquette.server.netty.NettyMQTTHandler.channelRead(NettyMQTTHandler.java:61)
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:318)
    at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:304)
    at io.moquette.server.netty.metrics.MessageMetricsHandler.channelRead(MessageMetricsHandler.java:46)
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:318)
    at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:304)
    at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:276)
    at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:263)
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:318)
    at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:304)
    at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:86)
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:318)
    at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:304)
    at io.moquette.server.netty.metrics.BytesMetricsHandler.channelRead(BytesMetricsHandler.java:47)
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:318)
    at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:304)
    at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:266)
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:318)
    at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:304)
    at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:846)
    at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:131)
    at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:511)
    at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:468)
    at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:382)
    at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:354)
    at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:112)
    at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:137)
    at java.lang.Thread.run(Thread.java:745)
```

---

Using the below code and org.eclipse.paho:org.eclipse.paho.client.mqttv3:1.0.2.
The payload is being received just fine, but the message appears in the log on the server. I have another client mqtt spy that works just fine. 

String client_id = MqttClient.generateClientId();

```
     MqttConnectOptions options = new MqttConnectOptions();

     MqttClient         client  = new MqttClient( brokerUrl,  // URI 
                                                  client_id,                      // ClientId 
                                                  new MemoryPersistence());       // Persistence

     options.setWill( sub_topic_err, gson.toJson(new Message( client_id , "Crashed!")).getBytes(), 2, true);

     client.setCallback( this );         

     try 
     {
         do {

           if( !client.isConnected() )
           {
             client.connect( options );
           }

           log.info("Subscribing to topic {}", sub_topic );
           log.info("Waiting for message to process.....");

           client.subscribe(sub_topic, 1); // This is a blocking call

          // client.disconnect();

           try{Thread.sleep(1000*maxRetrySleep);}catch(InterruptedException e){log.error(e.getMessage());}

           log.info("Going to sleep for {} seconds", maxRetrySleep);

         } while( true );
     }
     finally 
     {
          client.disconnect();
     }
```

**Comment by kjayashankar** [comment]
Coincidentally, it happens only if messageID = 2 (sometimes) and if I see too many of these on broker console, it gives #208

**Comment by nexusz99** [comment]
I have a same error. I published 200+ message on A topic and subscribed at the same time by 2 clients. And restarting both clients then i got a that error message on broker. Sometimes, subscribe didn't work..

**Comment by paulmiu** [comment]
@kfowlks: Exactly the same for me!
@kjayashankar: Do you mean messageID = 2 or QoS = 2?

**Comment by exxoff** [comment]
I get this error if I connect a client, kill the moquette broker and reconnect the client without resubscribing to any topics and then tries to publish a message.
If I do a Subscribe() prior to to Connect() it works but I get multiple subscriptions in the broker
```
1114 [nioEventLoopGroup-3-1] INFO  ProtocolProcessor  - CONNECT processed

9251 [nioEventLoopGroup-3-2] INFO  ClientSession  - cleaning old saved subscriptions for client Client1>
9274 [nioEventLoopGroup-3-2] INFO  ProtocolProcessor  - Connected client ID <Client1> with clean session true

9275 [nioEventLoopGroup-3-2] INFO  ProtocolProcessor  - CONNECT processed

9277 [nioEventLoopGroup-3-2] INFO  ClientSession  - <Client1> subscribed to topicFilter <Client1> with QoS 1 - LEAST_ONE

9290 [nioEventLoopGroup-3-2] INFO  ClientSession  - <Client1> subscribed to topicFilter <All> with QoS 1 - LEAST_ONE

9294 [nioEventLoopGroup-3-2] INFO  ClientSession  - <Client1> subscribed to topicFilter <Client1> with QoS 1 - LEAST_ONE

9295 [nioEventLoopGroup-3-2] INFO  ClientSession  - <Client1> subscribed to topicFilter <All> with QoS 1 - LEAST_ONE

9297 [nioEventLoopGroup-3-2] INFO  ClientSession  - <Client1> subscribed to topicFilter <Client1> with QoS 1 - LEAST_ONE

9298 [nioEventLoopGroup-3-2] INFO  ClientSession  - <Client1> subscribed to topicFilter <All> with QoS 1 - LEAST_ONE
```

To go around this I have to do Unsubscribe() and then subscribe to the topics again.

**Comment by andsel** [comment]
Which broker version are you using?
 Andrea

**Comment by exxoff** [comment]
Version 0.8. The jar is built Jan 17. I know that there's a 0.9 but I'm not able to build it on my current system. 0.8.1 was the latest distribution released.

**Comment by andsel** [comment]
No the latest are 0.9 and the current master is on 0.10-SNAPSHOT
Plase check with 0.9 or current master.
 Andrea

**Comment by exxoff** [comment]
Edit: Found it. The link in the readme points to 0.8 but I tried to enter 0.9 in the link instead and it worked.

https://bintray.com/artifact/download/andsel/generic/distribution-0.9-bundle-tar.tar.gz

---
I realize that there are other versions but they are not built and I don't have the possibility to build it.

If you could build it into jars, I'd be happy to test it.

 / Per

**Comment by exxoff** [comment]
The issue I had seems to be fixed in 0.9. Thanks!

 / Per

**Comment by kfowlks** [comment]
Based on the response from @exxoff , I'll go ahead and close this issue.
